File: 2444-count-subarrays-with-fixed-bounds/2444-count-subarrays-with-fixed-bounds.ts
function countSubarrays(nums: number[], minK: number, maxK: number): number {
    // Sliding Window
    // Time Complexity O(n)
    // Space Complexity O(1)

    let count = 0;
    let minKIdx = -1, maxKIdx = -1, outOfBoundsIdx = -1;

    for (let i = 0; i < nums.length; i++) {
        const v = nums[i];

        if (v < minK || v > maxK) {
            outOfBoundsIdx = i;
        }

        if (v === minK) {
            minKIdx = i;
        }

        if (v === maxK) {
            maxKIdx = i;
        }

        const validStart = Math.min(minKIdx, maxKIdx);
        count += Math.max(0, validStart - outOfBoundsIdx);
    }

    return count;
};

File: 2444-count-subarrays-with-fixed-bounds/2444-count-subarrays-with-fixed-bounds.py
class Solution:
    def countSubarrays(self, nums: List[int], minK: int, maxK: int) -> int:
        # Sliding Window
        # Time Complexity O(n)
        # Space Complexity O(1)
        count = 0

        minKIdx = maxKIdx = outOfBoundsIdx = -1

        for i, v in enumerate(nums):
            if v < minK or v > maxK:
                outOfBoundsIdx = i
            if v == minK:
                minKIdx = i
            if v == maxK:
                maxKIdx = i
            
            validStart = min(minKIdx, maxKIdx)
            count += max(0, validStart - outOfBoundsIdx)

        return count
        